bwt/wavelet.go

```diff
@@ -0,0 +1,381 @@
+as one example. Please reference the implementation if you'd like to understand how this
+specific waveletTree works.
+
+*/
+
+// waveletTree is a data structure that allows us to index a sequence
+// in a memory efficient way that allows us to conduct RSA, (R)ank (S)elect (A)ccess
+// queries on strings. This is very useful in situations where you'd like to understand
+// certain aspects of a sequence like:
+// * the number of times a character appears
+// * counting how the frequency of a character up to certain offset
+// * locating characters of certain rank within the sequence
+// * accessing the character at a given position
+type waveletTree struct {
+	root  *node
+	alpha []charInfo
+}
+
+// Access will return the ith character of the original
+// string used to build the waveletTree
+func (wt waveletTree) Access(i int) byte {
+	curr := wt.root
+	for !curr.isLeaf() {
+		bit := curr.data.Access(i)
+		i = curr.data.Rank(bit, i)
+		if bit {
+			curr = curr.right
+		} else {
+			curr = curr.left
+		}
+	}
+	return curr.char
+}
+
+// Rank allows us to get the rank of a specified character in
+// the original string
+func (wt waveletTree) Rank(char byte, i int) int {
+	curr := wt.root
+	ci := wt.lookupCharInfo(char)
+	level := 0
+	var rank int
+	for !curr.isLeaf() {
+		pathBit := ci.path.getBit(ci.path.len() - 1 - level)
+		rank = curr.data.Rank(pathBit, i)
+		if pathBit {
+			curr = curr.right
+		} else {
+			curr = curr.left
+		}
+		level++
+		i = rank
+	}
+	return rank
+}
+
+// Select allows us to get the corresponding position of a character
+// in the original string given its rank.
+func (wt waveletTree) Select(char byte, rank int) int {
+	curr := wt.root
+	ci := wt.lookupCharInfo(char)
+	level := 0
+
+	for !curr.isLeaf() {
+		pathBit := ci.path.getBit(ci.path.len() - 1 - level)
+		if pathBit {
+			curr = curr.right
+		} else {
+			curr = curr.left
+		}
+		level++
+	}
+
+	for curr.parent != nil {
+		curr = curr.parent
+		level--
+		pathBit := ci.path.getBit(ci.path.len() - 1 - level)
+		nextRank, ok := curr.data.Select(pathBit, rank)
+		if !ok {
+			msg := fmt.Sprintf("could not find a corresponding bit for node.Select(%t, %d) for characterInfo %+v", pathBit, rank, ci)
+			panic(msg)
+		}
+		rank = nextRank
+	}
+
+	return rank
+}
+
+func (wt waveletTree) lookupCharInfo(char byte) charInfo {
+	for i := range wt.alpha {
+		if wt.alpha[i].char == char {
+			return wt.alpha[i]
+		}
+	}
+	msg := fmt.Sprintf("could not find character %s in alphabet %+v. this should not be possible and indicates that the WaveletTree is malformed", string(char), wt.alpha)
+	panic(msg)
+}
+
+type node struct {
+	data   rsaBitVector
+	char   byte
+	parent *node
+	left   *node
+	right  *node
+}
+
+func (n node) isLeaf() bool {
+	return n.char != 0
+}
+
+type charInfo struct {
+	char    byte
+	maxRank int
+	path    bitvector
+}
+
+func NewWaveletTreeFromString(str string) waveletTree {
```

Comment:
is this supposed to be public? Needs Doc string

Reply:
It is not. ty. Made it private.